components/patient-data-sharing/receiver-api/src/main/java/org/phenotips/data/receive/internal/DefaultReceivePatientData.java

```diff
@@ -508,6 +538,43 @@ public JSONObject receivePatient()
         }
     }
 
+    /**
+     * Exctacts the list of granted consents from a request
+     * @param rawPatientState patient state JSON string directly from the {@link Request} object
+     */
+    private List<String> extractConsents(String rawPatientState)
+    {
+        List<String> consents = new LinkedList<>();
+        JSONObject patientState = this.patientStateToJson(rawPatientState);
+        if (patientState != null) {
+            try {
+                JSONArray consentsJson =
+                    patientState.getJSONArray(ShareProtocol.CLIENT_POST_KEY_NAME_PATIENTSTATE_CONSENTS);
+                for (Object consent : consentsJson) {
+                    consents.add(consent.toString());
+                }
+            } catch (Exception ex) {
+                // do nothing
+            }
+        }
+        return consents;
+    }
+
+    /**
+     * Converts raw patient state string from the request to {@link JSONObject}, or {@link null} if the function fails.
+     */
+    private JSONObject patientStateToJson(String rawPatientState)
+    {
+        if (rawPatientState != null) {
+            try {
+                return JSONObject.fromObject(URLDecoder.decode(rawPatientState, XWiki.DEFAULT_ENCODING));
+            } catch (Exception ex) {
```

Comment:
do we want to catch all exceptions? This also catches runtime exceptions..

If yes, then technically we don't need to do the null check, since it'll catch the null pointer exception :P